wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java

```diff
@@ -58,293 +59,187 @@
-                          "with predefined name and password",
-                  response = UserDescriptor.class)
-    @ApiResponses({@ApiResponse(code = 201, message = "Created"),
+    @GenerateLink(rel = LINK_REL_USER)
+    @ApiOperation(value = "Create a new user", response = UserDto.class)
+    @ApiResponses({@ApiResponse(code = 201, message = "User successfully created, response contains created entity"),
                    @ApiResponse(code = 400, message = "Missed required parameters, parameters are not valid"),
                    @ApiResponse(code = 401, message = "Missed token parameter"),
-                   @ApiResponse(code = 403, message = "Invalid or missing request parameters"),
-                   @ApiResponse(code = 409, message = "Invalid token"),
-                   @ApiResponse(code = 500, message = "Internal Server Error")})
-    public Response create(@ApiParam(value = "New user")
-                           UserDescriptor userDescriptor,
-                           @ApiParam(value = "Authentication token")
+                   @ApiResponse(code = 500, message = "Couldn't create user due to internal server error")})
+    public Response create(@ApiParam("New user")
+                           UserDto userDto,
+                           @ApiParam("Authentication token")
                            @QueryParam("token")
                            String token,
-                           @ApiParam(value = "User type")
+                           @ApiParam("User type")
                            @QueryParam("temporary")
                            @DefaultValue("false")
                            Boolean isTemporary,
                            @Context
-                           SecurityContext context) throws ForbiddenException,
-                                                           BadRequestException,
+                           SecurityContext context) throws BadRequestException,
                                                            UnauthorizedException,
                                                            ConflictException,
-                                                           ServerException,
-                                                           NotFoundException {
-        if (!context.isUserInRole("system/admin") && !userSelfCreationAllowed) {
-            throw new ForbiddenException("Currently only admins can create accounts. Please contact our Admin Team for further info.");
+                                                           ServerException {
+        final User createdUser;
+        if (token == null) {
+            checkUser(userDto);
+            createdUser = userManager.create(userDto, isTemporary);
+        } else {
+            if (!userSelfCreationAllowed) {
+                throw new BadRequestException("User self creation is not allowed");
+            }
+            createdUser = userManager.create(fromToken(token), isTemporary);
```

Comment:
Well, i think it is not a good approach.
- `token` and `userDto` are mutual exclusive
-  we don't always have session user instance in the manager, while we always do in the service
which excludes `userSelfCreationAllowed` check from the manager
- token and userDto should be valid, _bean valid_ and all the bean validations are performed in the service, i agree that we can extract this validation to the validators but, it is still service level
- Imho manager should keep clear and context independent operations which may be called from any place + contain all the business logic checks and operations, like `UserManager` creates profile along with a user when you call `.create(User user)`


What if in the other place we need a different parameter should manager be extended with a new create operation? I think no, while we can extend rest service with a new operations/parameters we should do so with manager